### PR TITLE
Oracle: download tokens

### LINKED ISSUE #915: Download tokens.xml on Oracle update
Sugestion time, it would be cool if Oracle would download tokens.xml from https://github.com/Psithief/Cockatrice-tokens to the default directory C:\Users\Rui Miranda\AppData\Local\Cockatrice\Cockatrice\tokens.xml

I'll play around to see what I can do, not much experience with c++, but will give it a try. Any tips are welcome!

### PR DESCRIPTION
**Pull request by ctrlaltca**
Fix #915.

After downloading cards and saving them, oracle will now download @Psithief 's token database and save it, too.
I suggest (if Psithief agrees to this change):
 1. creating a new "tokens" repo inside the cockatrice org;
 2. give Psithief commit access to it;
 3. change the default url from his own repo to cockatrice's repo.

### CODE REVIEW

**Comment by Nightfirecat** [comment]
:+1:

File: oracle/src/oraclewizard.cpp [hunk 1]
```diff
@@ -100,6 +104,17 @@ void OracleWizard::disableButtons()
...
 }
 
+bool OracleWizard::saveTokensToFile(const QString & fileName)
+{
+    QFile file(fileName);
+    if(!file.open(QIODevice::WriteOnly))
+        return false;
+    if(file.write(tokensData) == -1)
+        return false;
+    file.close();
```
Comment (Daenyth) [inline oracle/src/oraclewizard.cpp]: Log failures?

File: oracle/src/oraclewizard.cpp [hunk 2]
```diff
@@ -529,3 +544,203 @@ bool SaveSetsPage::validatePage()
...
+
+    wizard()->disableButtons();
+    setEnabled(false);
+
+    if(!nam)
+        nam = new QNetworkAccessManager(this);
+    QNetworkReply *reply = nam->get(QNetworkRequest(url));
+
+    connect(reply, SIGNAL(finished()), this, SLOT(actDownloadFinishedTokensFile()));
+    connect(reply, SIGNAL(downloadProgress(qint64, qint64)), this, SLOT(actDownloadProgressTokensFile(qint64, qint64)));
```
Comment (Daenyth) [inline oracle/src/oraclewizard.cpp]: Isn't there a failure signal too?

File: oracle/src/oraclewizard.cpp [hunk 3]
```diff
@@ -529,3 +544,203 @@ bool SaveSetsPage::validatePage()
...
+}
+
+void LoadTokensPage::actDownloadProgressTokensFile(qint64 received, qint64 total)
+{
+    if(total > 0)
+    {
+        progressBar->setMaximum(total);
+        progressBar->setValue(received);
+    }
+    progressLabel->setText(tr("Downloading (%1MB)").arg((int) received / 1048576));
```
Comment (Daenyth) [inline oracle/src/oraclewizard.cpp]: What's the hardcoded number here?

File: oracle/src/oraclewizard.cpp [hunk 4]
```diff
@@ -529,3 +544,203 @@ bool SaveSetsPage::validatePage()
...
+        QMessageBox::critical(this, tr("Error"), tr("Network error: %1.").arg(reply->errorString()));
+
+        wizard()->enableButtons();
+        setEnabled(true);
+
+        reply->deleteLater();
+        return;
+    }
+
+    // save allsets.json url, but only if the user customized it and download was successfull
```
Comment (Daenyth) [inline oracle/src/oraclewizard.cpp]: bad comment?

File: oracle/src/oraclewizard.cpp [hunk 5]
```diff
@@ -529,3 +544,203 @@ bool SaveSetsPage::validatePage()
...
+}
+
+bool SaveTokensPage::validatePage()
+{
+    bool ok = false;
+    const QString dataDir = 
+#if QT_VERSION < 0x050000
+        QDesktopServices::storageLocation(QDesktopServices::DataLocation);
+#else
+        QStandardPaths::standardLocations(QStandardPaths::DataLocation).first();
```
Comment (Daenyth) [inline oracle/src/oraclewizard.cpp]: We should make a util function in one location that does the ifdef check internally so we don't need to keep repeating it.

Doesn't need to happen for this pr

**Comment by woogerboy21** [comment]
Aside from the comments made by @Daenyth , the actual code for downloading the tokens file works.  The default URL is ugly but that was mentioned as a recommendation in the original post for adding a repo location.

File: oracle/src/oraclewizard.cpp (hunk 3 above)
Comment (Nightfirecat) [inline oracle/src/oraclewizard.cpp]: One megabyte, if calculated as 2^20 bytes.

File: oracle/src/oraclewizard.cpp (hunk 2 above)
Comment (ctrlaltca) [inline oracle/src/oraclewizard.cpp]: error() is emitted when an error occurs. finished() is emitted when the request ends, both if it succeeded or it failed.

File: oracle/src/oraclewizard.cpp (hunk 3 above)
Comment (ctrlaltca) [inline oracle/src/oraclewizard.cpp]: yup, we are receiving a qint64 number of bytes, / (1024 * 1024) = MB

**Comment by Psithief** [comment]
> I suggest (if Psithief agrees to this change):

Yes.

File: oracle/src/oraclewizard.cpp (hunk 4 above)
Comment (ctrlaltca) [inline oracle/src/oraclewizard.cpp]: indeed

**Comment by ctrlaltca** [comment]
> I suggest (if Psithief agrees to this change):
> Yes.

@Daenyth if you agree, too, can you create the new repo and set the appropriate access permissions?

File: oracle/src/oraclewizard.cpp (hunk 3 above)
Comment (Daenyth) [inline oracle/src/oraclewizard.cpp]: Can you expand this out to `1024 * 1024` to make that more clear? the compiler will emit the same output for it

**Comment by Daenyth** [comment]
Will do

**Comment by ZeldaZach** [comment]
Tested and works fully as expected. :shipit:

**Pull request merged**